### ISSUE TITLE
Improve speed of Docker container on (M+) macOS (and Windows?) 

### ISSUE THREAD

**Issue by hturner** [r-devel/r-dev-env#112]
Using the Docker container locally on an M1 mac is painfully slow, whether you launch from clone of the repo or a copy extracted from zip.

Approx timings:

| Action | Local | Codespaces |
|-------|-------|----------------|
| Configure R | 5-10 min | 1 min |
| Build R | > 35 min* | 5-10 min |

*build did not complete, I forget why and not inclined to repeat experiment!

There is some discussion about this issue [on Reddit](https://www.reddit.com/r/vscode/comments/14osu8u/devcontainer_in_apple_silicon_is_really_slow/) and [in the VSCode docs](https://code.visualstudio.com/remote/advancedcontainers/improve-performance). 

As our focus is on the codespace which is working well, this is low priority to fix, I am mainly creating this issue for reference.

For now we can put a note on the local setup documentation that its is not recommended to run the container locally on macOS or Windows.

**Comment by iaine** [comment]
I was looking at another issue and chanced on this message in Docker Desktop which might shed some light on this issue:
<img width="579" alt="Screenshot 2024-08-16 at 09 33 52" src="https://github.com/user-attachments/assets/638128bc-6058-4787-a3a2-e9be67f35a35">

There's a Docker forum answer about it here: https://forums.docker.com/t/run-x86-intel-and-arm-based-images-on-apple-silicon-m1-macs/117123 and also there do seem to be blog/medium posts (though some use other tools as well)

**Comment by bobturneruk** [comment]
Hi! We (me, @ubdbra001, @annakrystalli) tried a few changes to the Dockerfile to try and get it to build on ARM. This may not be new information, but we found that the `r-base-dev` installation step was problematic, with errors like:

```
#0 994.0  r-base-dev : Depends: r-base-core (>= 4.4.1-1.2204.0) but it is not going to be installed
```

**Comment by ubdbra001** [comment]
Just to add a bit more info:
- We identified the initial sticking point as lines 13-21 in the dockerfile and broke it down to work out what was going on
- Line 19 (`apt -y install r-base-dev`) throws the first error, as it needs `pkgconf` but is not installed for some reason.
- Adding a line to install `pkgconf` then produces the error @bobturneruk included above
- Installing `r-base-core` manually doesn't work at it doesn't look like there is a recent arm build for it (the version installed by running `apt install r-base-core` is `4.1.2-1ubuntu2`)

I don't think there are arm binaries for R through the [ubuntu r-project repo](https://cloud.r-project.org/bin/linux/ubuntu/), but there may be for [debian](https://cran.r-project.org/bin/linux/debian/#supported-packages). Using package source instead may solve the issues we found.

**Comment by hturner** [comment]
Thanks for this analysis, which was very informative. Searching for the error message posted by @bobturneruk led me to this [Stackoverflow Q & A](https://stackoverflow.com/q/78228725/173755), which in turn led me to look at the [rocker/r-ubuntu Dockerfile](https://github.com/rocker-org/rocker/blob/master/r-ubuntu/focal/Dockerfile#L28). There they use "ppa:marutter/rrutter4.0" as the repository vs "https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/" as we have been using. I tried building from the [R Dev Container Dockerfile](https://github.com/r-devel/r-dev-env/blob/main/Dockerfile) with Michael Rutter's base R PPA as the R repository and the build was successful on my ARM machine!

PR #197 updates the Dockerfiles with this change. I think we will now be able to get this working with buildx. Looking at the [buildx docs](https://github.com/marketplace/actions/docker-setup-buildx) it looks like we only need to set up Docker Buildx and not QEMU. @StarTrooper08 do you want to have a go at this?

**Comment by iaine** [comment]
Have just run a test on an M1 from 2020 to build R in a container using a script here: https://github.com/iaine/time_r  
I have it largely building in 6 minutes 29 seconds when it crashes running:

```Testing examples for package ‘tcltk’
Error: running examples in ‘tcltk-Ex.R’ failed
Error: testing 'tcltk' failed
Execution halted
make[3]: *** [Makefile:37: test-Examples-Base] Error 1
make[3]: Leaving directory '/workspaces/r-dev-env/build/r-devel/tests/Examples'
make[2]: *** [/workspaces/r-dev-env/svn/r-devel/tests/Makefile.common:216: test-Examples] Error 2
make[2]: Leaving directory '/workspaces/r-dev-env/build/r-devel/tests'
make[1]: *** [/workspaces/r-dev-env/svn/r-devel/tests/Makefile.common:201: test-all-basics] Error 1
make[1]: Leaving directory '/workspaces/r-dev-env/build/r-devel/tests'
make: *** [Makefile:310: check] Error 2```

**Comment by iaine** [comment]
I managed to get it building R in a container on an M1 OSX machines from 2020 with two runs. The times, including all tests, were:

0h:30m:42s
0h:28m:42s

So not fast. The container creation was relatively quick.

**Comment by iaine** [comment]
I rebuilt an linux/arm64 container using docker buildx on an M1 Mac using the time_r script from above with the time:
build rsync-recommended -> 0h:0m:4s
configure -> 0h:0m:13s
make -> 0h:5m:50s
make check -> 0h:1m:56s
0h:8m:12s

I got similar results across both runs. I have not run the linux/amd64 yet as  I'm on an arm machine and suspect that this was behind the slower time earlier.

**Comment by hturner** [comment]
As discussed on Slack, we plan to switch to a reusable workflow, to avoid duplicating code and make it easier to test options for building the arm64 images. We can do this step-by-step, as set-out in sub-issues #252, #253, #254.

Possibly we could go straight to #254 after #252, since the buildx setup is a little more complicated. But it could be good to get both working, in case we face issues with the linux arm64 runners.
